pkg/controller/metadata/metadata_controller_test.go

```diff
@@ -558,3 +558,38 @@ func generateCertChain(t *testing.T, ctx context.Context, c client.Client, g *Go
 
 	return rootCertReq, intCertReq, metaCertReq, tearDown
 }
+
+func TestShouldRegenerate(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	const ConstantHash = "Im a constant hash"
+
+	mockSecrets := corev1.Secret{}
+	mockSecrets.ObjectMeta.Annotations = make(map[string]string)
+	mockSecrets.Data = make(map[string][]byte)
+
+	// Hashes should differ, so should be true to regenerate
+	mockSecrets.ObjectMeta.Annotations[VersionAnnotation] = ""
+	g.Eventually(ShouldRegenerate(&mockSecrets, ConstantHash)).Should(BeTrue())
+
+	// Hash should now match, but there is no data for the expiration, this simulates a upgrade.
+	mockSecrets.ObjectMeta.Annotations[VersionAnnotation] = ConstantHash
+	g.Eventually(ShouldRegenerate(&mockSecrets, ConstantHash)).Should(BeTrue())
+
+	// There should be a parse error.
+	mockSecrets.Data[ValidityDays] = []byte("30")
+	mockSecrets.Data[ValidUntil] = []byte("")
+	g.Eventually(ShouldRegenerate(&mockSecrets, ConstantHash)).Should(BeTrue())
+
+	// Should regenerate as time is in the past.
+	mockSecrets.Data[ValidUntil] = []byte("Mon, 02 Jan 2006 15:04:05 -0700")
+	g.Eventually(ShouldRegenerate(&mockSecrets, ConstantHash)).Should(BeTrue())
+
+	// Should regenerate if half of the validity days.
+	mockSecrets.Data[ValidUntil] = []byte(time.Now().AddDate(0, 0, 14).Format(time.RFC1123Z))
+	g.Eventually(ShouldRegenerate(&mockSecrets, ConstantHash)).Should(BeTrue())
+
+	// Should'nt regenerate as in the future.
+	mockSecrets.Data[ValidUntil] = []byte(time.Now().AddDate(0, 0, 60).Format(time.RFC1123Z))
```

Comment:
Is this adding 60 days to `time.Now()`? This will work provided we never set `validUntil` to more than 1119 days - which is unlikely but maybe it would be better to do something like `time.Now() + 2*validityDays + 1` to make sure the threshold is never reached